src/network/network_survey.cpp

```diff
@@ -0,0 +1,356 @@
+ * See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with OpenTTD. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/** @file network_survey.cpp Opt-in survey part of the network protocol. */
+
+#include "../stdafx.h"
+#include "network_survey.h"
+#include "settings_table.h"
+#include "network.h"
+#include "../debug.h"
+#include "../rev.h"
+#include "../settings_type.h"
+#include "../timer/timer_game_tick.h"
+
+#include "../currency.h"
+#include "../fontcache.h"
+#include "../language.h"
+
+#include "../ai/ai_info.hpp"
+#include "../game/game.hpp"
+#include "../game/game_info.hpp"
+
+#include "../music/music_driver.hpp"
+#include "../sound/sound_driver.hpp"
+#include "../video/video_driver.hpp"
+
+#include "../base_media_base.h"
+#include "../blitter/factory.hpp"
+
+#ifdef WITH_NLOHMANN_JSON
+#include <nlohmann/json.hpp>
+#endif /* WITH_NLOHMANN_JSON */
+
+#include "../safeguards.h"
+
+extern std::string _savegame_uuid;
+
+NetworkSurveyHandler _survey = {};
+
+#ifdef WITH_NLOHMANN_JSON
+
+NLOHMANN_JSON_SERIALIZE_ENUM(NetworkSurveyHandler::Reason, {
+	{NetworkSurveyHandler::Reason::LEAVE, "leave"},
+	{NetworkSurveyHandler::Reason::EXIT, "exit"},
+	{NetworkSurveyHandler::Reason::CRASH, "crash"},
+})
+
+NLOHMANN_JSON_SERIALIZE_ENUM(GRFStatus, {
+	{GRFStatus::GCS_UNKNOWN, "unknown"},
+	{GRFStatus::GCS_DISABLED, "disabled"},
+	{GRFStatus::GCS_NOT_FOUND, "not found"},
+	{GRFStatus::GCS_INITIALISED, "initialised"},
+	{GRFStatus::GCS_ACTIVATED, "activated"},
+})
+
+static const std::string _vehicle_type_to_string[] = {
+	"train",
+	"roadveh",
+	"ship",
+	"aircraft",
+};
+
+/* Defined in one of the os/ survey files. */
+extern void SurveyOS(nlohmann::json &json);
+
+/**
+ * List of all the generic setting tables.
+ *
+ * There are a few tables that are special and not processed like the rest:
+ * - _currency_settings
+ * - _misc_settings
+ * - _company_settings
+ * - _win32_settings
+ * As such, they are not part of this list.
+ */
+static auto &GenericSettingTables()
+{
+	static const SettingTable _generic_setting_tables[] = {
+		_difficulty_settings,
+		_economy_settings,
+		_game_settings,
+		_gui_settings,
+		_linkgraph_settings,
+		_locale_settings,
+		_multimedia_settings,
+		_network_settings,
+		_news_display_settings,
+		_pathfinding_settings,
+		_script_settings,
+		_world_settings,
+	};
+	return _generic_setting_tables;
+}
+
+/**
+ * Convert a settings table to JSON.
+ *
+ * @param survey The JSON object to add the settings to.
+ * @param table The settings table to convert.
+ * @param object The object to get the settings from.
+ */
+static void SettingsTableToJson(nlohmann::json &survey, const SettingTable &table, void *object)
+{
+	char buf[512];
+	for (auto &desc : table) {
+		const SettingDesc *sd = GetSettingDesc(desc);
+		/* Skip any old settings we no longer save/load. */
+		if (!SlIsObjectCurrentlyValid(sd->save.version_from, sd->save.version_to)) continue;
```

Comment:
Maybe also skip settings not different than default value ?

Reply:
That would mean we need to keep a backend-side store of default values; I see what you mean, and it would save (a lot of) storage, but it would also complicate the flow of information .. for every new release we also have to tell how settings are doing etc.

Meh. Both have their merits :)

Reply:
Gave this some thought, and I rather have a maintenance free system, with as little moving parts as I can. So, I think we should go for sending all settings :)

We compress survey results in the backend, which is doing an amazing job (1:100 in early benchmarks). So it is just a bit of bandwidth we waste; and with ~20KiB, that shouldn't be an issue.